### DISCUSSION TITLE
How can i add support for batch _entities in federation

### DISCUSSION THREAD

**Discussion by SaranSundar** [Netflix/dgs-framework#1488] [Q&A] [answered]
I'm converting a legacy app that was bare bones graphql java to netflix dgs

We do some wiring that looks like this 

```
type Employer @key(fields: "id") {
    id: Int!
    name: String
}

addFederatedEntityDataFetcher(Employer.class, query::findEmployersByIds)



@Override
    public List<Employer> findEmployersByIds(List<Map<String, Object>> parameters) {
        final List<Integer> employerIds = parameters.stream().map(param -> ((Number) param.get("id")).intValue()).collect(Collectors.toList());
        return employers.findByMultiId(employerIds);
    }
```

It's kind of similar to the dgs @DgsEntityFetcher, except instead of being a singular entity fetcher, it allows performant batching on the app side to load hundreds of employers with a single db query vs resolving one at a time.

I see there is a @DgsDataLoader, but im confused how these work in conjunction. Is a DgsEntityFetcher supposed to somehow call a DgsDataLoader? or how is a DgsDataLoader suposed to be used. And does dgs support the above of being able to efficiently batch the entities fetcher for type Employer?

**Comment by srinivasankavitha** [comment]
You can set up a DgsDataLoader and use that within an data fetcher (or entity fetcher) like this: https://netflix.github.io/dgs/data-loaders/#using-a-data-loader

Hope this helps.

**Reply by SaranSundar** [reply]
So for my case Employer is the parent type, would i replace dgs data with DgsEntityFetcher  where the final output would look like

```
@DgsComponent
public class EmployerEntityFetcher {

    @DgsEntityFetcher(name = "Employer")
    public CompletableFuture<Employer> employer(DataFetchingEnvironment dfe) {

        DataLoader<Integer, Employer> dataLoader = dfe.getDataLoader("employers");
        Integer id = dfe.getArgument("id");

        return dataLoader.load(id);
    }
}
```

```
@DgsDataLoader(name = "employers")
public class EmployersDataLoader implements BatchLoader<Integer, Employer> {

    @Autowired
    EmployerServiceClient employerServiceClient;

    @Override
    public CompletionStage<List<Employer>> load(List<Integer> keys) {
        return CompletableFuture.supplyAsync(() -> employerServiceClient.loadEmployers(keys));
    }
}
```

Given schema like
```
type Employer @key(fields: "id") {
    id: Int!
    name: String
}
```

**Reply by SaranSundar** [reply]
@srinivasankavitha does the above code look right for allowing batching of entities of the Employer parent type?

**Comment by srinivasankavitha** [comment]
Yes, if Employer is the type that is requested as part of the _entities
query, that looks correct to me.

**Reply by SaranSundar** [reply, marked as answer]
@srinivasankavitha Thanks I got a federated apollo graph gateway running locally via passing in a service list to test out the entities, so a couple issues I noticed. One is that DgsEntityFetcher requires an argument of Map<String, Object> values which is fine. The real issue is the DataFetchingEnvironment dfe doesnt seem to be populating the context.

In my normal queries I can grab context and it populates the variables passed from headers, but when testing the federation its not passing the context through to the entity fetcher.

`final Context context = DgsContext.getCustomContext(env);`

```
  @DgsEntityFetcher(name = "Employer")
  public CompletableFuture<Employer> employer(Map<String, Object> values, DataFetchingEnvironment dfe) {

    DataLoader<Integer, Employer> dataLoader = dfe.getDataLoader("employer");
    Integer id = (Integer) values.get("id");

    return dataLoader.load(id);
  }
```

**Reply by SaranSundar** [reply]
Also is there a way to get the selectionSet from a BatchLoaderEnvironment, if not is there a way to get access to DataFetchingEnvironment in a DataLoader?

**Reply by srinivasankavitha** [reply]
You cannot access the DFE in the batch loader. However, you can pass in context using this mechanism : https://www.graphql-java.com/documentation/batching/#passing-context-to-your-data-loader. That is generally the way to hand over any context to data loaders. You can also use that as part of the key depending on your use case.

For the context in the entity fetcher, you should be able to extract the context from the data fetching environment inside the entity fetcher implementation via `DgsContext.getCustomContext(environment)` - the context is created and set per request. 

Hope that helps.

**Reply by SaranSundar** [reply]
Ok thanks i'll try that out!
